### PR TITLE
python3Packages.tuya-device-sharing-sdk: 0.2.0 -> 0.1.9

### LINKED ISSUE #319067: python3Packages.tuya-device-sharing-sdk: Discuss downgrade to version 0.1.9
Hi @aciceri @mweinelt

I would like to discuss the following issue with you.

## Issue description

The python library tuya-device-sharing-sdk is currently packaged as version 0.2.0 [(link)](https://github.com/NixOS/nixpkgs/blob/8f5f49d8ad22689611fcbb62b1a79124a35cf67d/pkgs/development/python-modules/tuya-device-sharing-sdk/default.nix#L12), and used only by the Home Assistant component "Tuya". This works because of the way nixpkgs is currently "allocating" dependencies for official Home assistant components [(link)](https://github.com/NixOS/nixpkgs/blob/8f5f49d8ad22689611fcbb62b1a79124a35cf67d/pkgs/servers/home-assistant/component-packages.nix#L4716), totally ignoring version constraints of the component itself [(link)](https://github.com/search?q=repo%3Ahome-assistant%2Fcore%20tuya-device-sharing-sdk&type=code).

I have packaged the custom component tuya-local with version 2024.5.2 for nixpks [(link)](https://github.com/NixOS/nixpkgs/blob/8f5f49d8ad22689611fcbb62b1a79124a35cf67d/pkgs/servers/home-assistant/custom-components/tuya_local/default.nix) that recently also added tuya-device-sharing-sdk as a dependency with a "hard constraint" on version 0.1.9 also [(link)](https://github.com/make-all/tuya-local/commit/8896547f6af5ea6f3252256e0dc94cc4e54aa983#diff-4d7c51b1efe9043e44439a949dfd92e5827321b34082903477fd04876edb7552) and the change got released with version 2024.6.0. As far as I know I cannot make nixpkgs build tuya-local with the new version 2024.6.0 because of the version mismatch of the tuya-device-sharing-sdk dependency, at least not without patching the version constraints.

But in general, we should probably ask whether we really want to provide version 0.2.0 of the tuya-device-sharing-sdk library in nixpkgs. This version got "released" from a dev branch [(link)](https://github.com/tuya/tuya-device-sharing-sdk/tree/dev) and the maintainers of both, the official Tuya component and the custom tuya-local components obviously don't want to use this version.

My suggestion would be to downgrade the version to version 0.1.9 that is actually requested by its dependents.

What do you think? I'm willing to provide the PR.

Thanks!

### PR DESCRIPTION
**Pull request by pathob**
## Description of changes

Closes https://github.com/NixOS/nixpkgs/issues/319067

TL;DR

Downgrade version from 0.2.0, which got released from a dev branch, to 0.1.9, that is actually required by it's only nixpkgs dependent; the Home Assistant "Tuya" component.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by pathob** [comment]
Since @aciceri mentioned that he is out-of-office, I would assign the PR to @mweinelt

I've run `nix-shell -p nixpkgs-review --run "nixpkgs-review pr 319093"` successfully, but it doesn't look like it's able to recognise that this library is used as a dependency in an official Home Assistant component. To me it seems that official Home Assistant components actually lack proper dependency handling compared to the custom Home Assistant components!?

I've also integrated my branch as nixpkgs-unstable into my flake. I'm not using the Tuya component, but everything builds as expected, there are no errors in Home Assistant and the integration also starts up without errors.

**Comment by pathob** [comment]
@mweinelt Can we merge this? I'd like to proceed with the other Tuya packages

**Comment by mweinelt** [comment]
Sorry, I merge the change in #319096, so I could test the impact while updating home-assistant.
